jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/BlockingTest.java

```diff
@@ -13,131 +13,102 @@
+                 runnable.run();
+             }))
         {
-            final CountDownLatch latch = new CountDownLatch(1);
-
-            new Thread(() ->
-            {
-                latch.countDown();
-                try
-                {
-                    TimeUnit.MILLISECONDS.sleep(100);
-                }
-                catch (Exception e)
-                {
-                    e.printStackTrace();
-                }
-                runnable.run();
-            }).start();
-
-            latch.await();
-            start = NanoTime.now();
+            thread.start();
             runnable.block();
         }
-        long elapsed = NanoTime.millisSince(start);
-        assertThat(elapsed, greaterThan(10L));
-        assertThat(elapsed, lessThan(1000L));
     }
 
     @Test
     public void testNoRun()
     {
-        long start;
         try (Blocker.Runnable ignored = Blocker.runnable())
         {
-            start = NanoTime.now();
             LoggerFactory.getLogger(Blocker.class).info("expect WARN Blocking.Runnable incomplete");
         }
-        assertThat(NanoTime.millisSince(start), lessThan(500L));
     }
 
     @Test
     public void testSucceededBlock() throws Exception
     {
-        long start;
         try (Blocker.Callback callback = Blocker.callback())
         {
             callback.succeeded();
-            start = NanoTime.now();
             callback.block();
         }
-        assertThat(NanoTime.millisSince(start), lessThan(500L));
     }
 
     @Test
     public void testBlockSucceeded() throws Exception
     {
-        long start;
-        try (Blocker.Callback callback = Blocker.callback())
+        try (Blocker.Callback callback = Blocker.callback();
+             AssertingThread thread = new AssertingThread(() ->
+             {
+                 await().atMost(5, TimeUnit.SECONDS).until(main::getState, Matchers.is(Thread.State.WAITING));
+                 callback.succeeded();
+             }))
         {
-            final CountDownLatch latch = new CountDownLatch(1);
-
-            new Thread(() ->
-            {
-                latch.countDown();
-                try
-                {
-                    TimeUnit.MILLISECONDS.sleep(100);
-                }
-                catch (Exception e)
-                {
-                    e.printStackTrace();
-                }
-                callback.succeeded();
-            }).start();
+            thread.start();
 
-            latch.await();
-            start = NanoTime.now();
             callback.block();
         }
-        long elapsed = NanoTime.millisSince(start);
-        assertThat(elapsed, greaterThan(10L));
-        assertThat(elapsed, lessThan(1000L));
     }
 
     @Test
-    public void testFailedBlock() throws Exception
+    public void testFailedBlock()
     {
-        final Exception ex = new Exception("FAILED");
-        long start = Long.MIN_VALUE;
+        Exception ex = new Exception("FAILED");
```

Comment:
Couldn't the entire `try(Blocker.Callback ..` block be inside an `assertThrows(Exception.class, () -> try ...)`?

Reply:
I'm guessing that it could look like ...

``` java
Exception ex = new Exception("FAILED");
IOException actual = assertThrows(IOException.class, () ->
    try (Blocker.Callback callback = Blocker.callback())
    {
        callback.failed(ex);
        callback.block();
    }
);
assertEquals(ex, actual.getCause());
```

Reply:
You're right, that more elegant and readable. Let me change this.